src/containers/tokens/components/bonding-curve-form/index.js

```diff
@@ -0,0 +1,422 @@
+import { connect } from 'react-redux'
+import { toBN } from 'ethjs'
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+import { formValueSelector } from 'redux-form'
+
+import Button from '../../../../components/button'
+import { form } from '../../../../utils/form-generator'
+import { required, number, positiveNumber } from '../../../../utils/validation'
+import {
+  decimalStringToWeiBN,
+  weiBNToDecimalString
+} from '../../../../utils/number'
+import * as bondingCurveSelectors from '../../../../reducers/bonding-curve'
+import * as bondingCurveActions from '../../../../actions/bonding-curve'
+
+const {
+  Form: BuyPNKFromBondingCurveForm,
+  isInvalid: getBuyPNKFromBondingCurveFormIsInvalid,
+  submit: submitBuyPNKFromBondingCurveForm
+} = form('buyPNKFromBondingCurveForm', {
+  header: {
+    type: 'header',
+    props: { title: 'BUY PNK' }
+  },
+  amountOfETH: {
+    type: 'text',
+    validate: [required, number, positiveNumber],
+    props: {
+      type: 'number',
+      className: 'Form-noMargins'
+    }
+  },
+  rate: {
+    type: 'info',
+    props: {
+      className: 'Tokens-form-rate'
+    }
+  }
+})
+
+const { Form: SellPNKToBondingCurveFormHeader } = form(
+  'sellPNKToBondingCurveFormHeader',
+  {
+    header: {
+      type: 'header',
+      props: { title: 'SELL PNK' }
+    }
+  }
+)
+
+const {
+  Form: SellPNKToBondingCurveForm,
+  isInvalid: getSellPNKToBondingCurveFormIsInvalid,
+  submit: submitSellPNKToBondingCurveForm
+} = form('sellPNKToBondingCurveForm', {
+  amountOfPNK: {
+    type: 'text',
+    validate: [required, number, positiveNumber],
+    props: {
+      type: 'number',
+      className: 'Form-noMargins'
+    }
+  },
+  rate: {
+    type: 'info',
+    props: {
+      className: 'Tokens-form-rate'
+    }
+  }
+})
+
+/**
+ * Convert a fraction to float.
+ * @param {BigNumber} numerator - The numerator.
+ * @param {BigNumber} denominator - The denominator.
+ * @returns {number} - The result.
+ */
+function toFloat(numerator, denominator) {
+  return (
+    numerator
+      .mul(toBN(10000))
+      .div(denominator)
+      .toNumber() / 10000
+  )
+}
+
+/**
+ * Round number to 4 significant figures.
+ * @param {string} n - Input number.
+ * @returns {string} - Rounded number with at least 4 significant figures.
+ */
+function round(n) {
+  // Pad the number so it has at least 4 significant figures.
+  if (n.indexOf('.') === -1) n += '.'
+
+  n += '0000'
+
+  var count = 0
+  var dot = false
+  for (var i = 0; i < n.length; i++) {
```

Comment:
No need for the extra loop

```
let decimalIndex = n.indexOf('.')
if (decimalIndex === -1) {
  n += '.' // add decimal if whole number
  decimalIndex = n.length - 1
} 
n += '0000' // add 4 extra decimal places
return n.slice(0, decimalIndex + 4)
```

Reply:
What I need is to round the number so that it has 4 significant figures, not 4 decimal places.

Reply:
The intention is to keep 4 significant digits but if the number >=10000 then keep the integral part. The reason for doing this instead of keeping 4 decimal places is so that 0.00012345 is truncated to 0.0001234 instead of 0.0001.

Reply:
Why do we need that much precision? For example for ETH the difference between 0.0001 and 0.0001234 is $0.004. That much precision makes the numbers harder to read and the amount is probably irrelevant to the user. We can (and should) make the number hoverable that shows the full amount but I think it is ok to truncate to fixed decimal places for display purposes depending on what we are displaying. ETH can be to 4 decimal places for example and PNK can be rounded to the nearest whole number

Reply:
OK. I will change as suggested. Hoverable is probably the best but I'll leave it for future improvement for now.